apps/android/LiveVideoApp/app/src/main/java/com/twilio/livevideo/app/manager/room/LocalParticipantWrapper.kt

```diff
@@ -0,0 +1,182 @@
+        if (!isAudioMuted) {
+            isAudioMuted = true
+            removeAudioTrack()
+        } else {
+            isAudioMuted = false
+            setupLocalAudioTrack()
+        }
+    }
+
+    fun toggleLocalVideo() {
+        if (!isVideoMuted) {
+            isVideoMuted = true
+            removeCameraTrack()
+        } else {
+            isVideoMuted = false
+            setupLocalVideoTrack()
+        }
+    }
+
+    private fun setupLocalAudioTrack() {
+        if (localAudioTrack == null && !isAudioMuted && context != null) {
+            localAudioTrack = createLocalAudioTrack(context, true, MICROPHONE_TRACK_NAME)
+            localAudioTrack?.let { publishAudioTrack(it) }
+                ?: Timber.e(RuntimeException(), "Failed to create local audio track")
+        }
+    }
+
+    private fun setupLocalVideoTrack() {
+        context?.let {
+            cameraCapturer = CameraCapturerCompat.newInstance(it)
+
+            videoTrack = cameraCapturer?.let { capturer ->
+                LocalVideoTrack.create(it, true, capturer, null, CAMERA_TRACK_NAME)
+            }?.apply {
+                localVideoTrackNames[name] = context.getString(R.string.camera_video_track)
+                publishCameraTrack(this)
+            }
+        }
+    }
+
+    private fun publishCameraTrack(localVideoTrack: LocalVideoTrack?) {
+        if (!isVideoMuted) {
+            localVideoTrack?.let {
+                participant?.publishTrack(
+                    it,
+                    LocalTrackPublicationOptions(TrackPriority.LOW)
+                )
+            }
+        }
+    }
+
+    private fun publishAudioTrack(localAudioTrack: LocalAudioTrack?) {
+        if (!isAudioMuted) {
+            localAudioTrack?.let { participant?.publishTrack(it) }
+        }
+    }
+
+    private fun unPublishTrack(localVideoTrack: LocalVideoTrack?) =
+        localVideoTrack?.let { participant?.unpublishTrack(it) }
+
+    private fun unPublishTrack(localAudioTrack: LocalAudioTrack?) =
+        localAudioTrack?.let { participant?.unpublishTrack(it) }
+
+    private fun removeCameraTrack() {
+        videoTrack?.let { cameraVideoTrack ->
+            unPublishTrack(cameraVideoTrack)
+            localVideoTrackNames.remove(cameraVideoTrack.name)
+            cameraVideoTrack.release()
+            this.videoTrack = null
+        }
+    }
+
+    private fun removeAudioTrack() {
+        localAudioTrack?.let { localAudioTrack ->
+            unPublishTrack(localAudioTrack)
+            localAudioTrack.release()
+            this.localAudioTrack = null
+        }
+    }
+
+    override fun onNetworkQualityLevelChanged(localParticipant: LocalParticipant, networkQualityLevel: NetworkQualityLevel) {
+        this.networkQualityLevel = networkQualityLevel
+    }
+
+    override fun onAudioTrackPublished(
+        localParticipant: LocalParticipant,
+        localAudioTrackPublication: LocalAudioTrackPublication
+    ) {
+        //TODO("Not yet implemented")
```

Comment:
I think for local participant this app only needs to handle the error callbacks. And so this happy path callback for example can probably just be removed.

Reply:
Yes, you right, I haven't figure it out, what methods are actually that we are going to use from this listener. However, in `Kotlin` and `Java`, as we implement an `Interface`, we should override/implement ALL the methods from that interface. In this case, we are forced to implement all the non-optional/non-default methods from the interface `LocalParticipant.Listener`

Reply:
Makes sense and seems better actually. That is the same with pure Swift, but because the SDK is Objective-C each callback is optional.